R/data.table.R

```diff
@@ -1337,8 +1337,22 @@ replace_dot_alias = function(e) {
           ## check key on i as well!
           ichk = is.data.table(i) && haskey(i) &&
                  identical(head(key(i), length(leftcols)), names_i[leftcols]) # i has the correct key, #3061
-          if (keylen && (ichk || is.logical(i) || (.Call(CisOrderedSubset, irows, nrow(x)) && ((roll == FALSE) || length(irows) == 1L)))) # see #1010. don't set key when i has no key, but irows is ordered and roll != FALSE
-            setattr(ans,"sorted",head(key(x),keylen))
+          dt_types = function(x, cols = seq_along(x)) vapply_1c(cols, function(j) typeof(x[[j]]))
```

Comment:
I guess the alternative is `vapply_1c(.shallow(i, leftcols), typeof)`, is that right? I'm not the most familiar with `.shallow()` magic :)

Reply:
Yes, see edit of benchmarks above. `.shallow` is even faster single only need to do a single shallow